go/vt/vtgate/engine/insert.go

```diff
@@ -0,0 +1,555 @@
+	Values []sqltypes.PlanValue
+
+	// Table sepcifies the table for the insert.
+	Table *vindexes.Table
+
+	// Generate is only set for inserts where a sequence must be generated.
+	Generate *Generate
+
+	// Prefix, Mid and Suffix are for sharded insert plans.
+	Prefix string
+	Mid    []string
+	Suffix string
+}
+
+// MarshalJSON serializes the Insert into a JSON representation.
+// It's used for testing and diagnostics.
+func (ins *Insert) MarshalJSON() ([]byte, error) {
+	var tname string
+	if ins.Table != nil {
+		tname = ins.Table.Name.String()
+	}
+	marshalInsert := struct {
+		Opcode   InsertOpcode
+		Keyspace *vindexes.Keyspace   `json:",omitempty"`
+		Query    string               `json:",omitempty"`
+		Values   []sqltypes.PlanValue `json:",omitempty"`
+		Table    string               `json:",omitempty"`
+		Generate *Generate            `json:",omitempty"`
+		Prefix   string               `json:",omitempty"`
+		Mid      []string             `json:",omitempty"`
+		Suffix   string               `json:",omitempty"`
+	}{
+		Opcode:   ins.Opcode,
+		Keyspace: ins.Keyspace,
+		Query:    ins.Query,
+		Values:   ins.Values,
+		Table:    tname,
+		Generate: ins.Generate,
+		Prefix:   ins.Prefix,
+		Mid:      ins.Mid,
+		Suffix:   ins.Suffix,
+	}
+	return jsonutil.MarshalNoEscape(marshalInsert)
+}
+
+// Generate represents the instruction to generate
+// a value from a sequence.
+type Generate struct {
+	Keyspace *vindexes.Keyspace
+	Query    string
+	// Values are the supplied values for the column, which
+	// will be stored as a list within the PlanValue. New
+	// values will be generated based on how many were not
+	// supplied (NULL).
+	Values sqltypes.PlanValue
+}
+
+// InsertOpcode is a number representing the opcode
+// for the Insert primitve.
+type InsertOpcode int
+
+const (
+	// InsertUnsharded is for routing an insert statement
+	// to an unsharded keyspace.
+	InsertUnsharded = InsertOpcode(iota)
+	// InsertSharded is for routing an insert statement
+	// to individual shards. Requires: A list of Values, one
+	// for each ColVindex. If the table has an Autoinc column,
+	// A Generate subplan must be created.
+	InsertSharded
+	// InsertShardedIgnore is for INSERT IGNORE and
+	// INSERT...ON DUPLICATE KEY constructs.
+	InsertShardedIgnore
+)
+
+var insName = map[InsertOpcode]string{
+	InsertUnsharded:     "InsertUnsharded",
+	InsertSharded:       "InsertSharded",
+	InsertShardedIgnore: "InsertShardedIgnore",
+}
+
+// MarshalJSON serializes the InsertOpcode as a JSON string.
+// It's used for testing and diagnostics.
+func (code InsertOpcode) MarshalJSON() ([]byte, error) {
+	return json.Marshal(insName[code])
+}
+
+// Execute performs a non-streaming exec.
+func (ins *Insert) Execute(vcursor VCursor, bindVars, joinVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	bindVars = combineVars(bindVars, joinVars)
```

Comment:
Unless I'm missing something here, `joinVars` should always be empty in this case and in the case of Update as well. Given that it seems better to assert as such rather than doing this unnecessary combineVars.

Reply:
Done.